src/rpc/server.zig

```diff
@@ -0,0 +1,547 @@
+
+            .snapshot_dir = params.snapshot_dir,
+            .latest_snapshot_gen_info = params.latest_snapshot_gen_info,
+
+            .wait_group = .{},
+            .thread_pool = params.thread_pool,
+
+            .read_buffer_size = @max(params.read_buffer_size, InitParams.MIN_READ_BUFFER_SIZE),
+            .tcp = tcp_server,
+        };
+    }
+
+    /// Blocks until all tasks are completed, and then closes the server.
+    /// Does not force the server to exit.
+    pub fn joinDeinit(server: *Server) void {
+        server.wait_group.wait();
+        server.tcp.deinit();
+    }
+
+    /// Spawn the serve loop as a separate thread.
+    pub fn serveSpawn(
+        server: *Server,
+        exit: *std.atomic.Value(bool),
+    ) std.Thread.SpawnError!std.Thread {
+        return std.Thread.spawn(.{}, serve, .{ server, exit });
+    }
+
+    /// Calls `acceptAndServeConnection` in a loop until `exit.load(.acquire)`.
+    pub fn serve(server: *Server, exit: *std.atomic.Value(bool)) !void {
+        while (!exit.load(.acquire)) {
+            _ = try server.acceptAndServeConnection();
+        }
+    }
+
+    pub const AcceptAndServeConnectionError =
+        std.mem.Allocator.Error ||
+        std.http.Server.ReceiveHeadError ||
+        AcceptConnectionError;
+
+    pub fn acceptAndServeConnection(server: *Server) AcceptAndServeConnectionError!void {
+        const conn = (try acceptConnection(&server.tcp, server.logger)).?;
+        errdefer conn.stream.close();
+
+        server.wait_group.start();
+        errdefer server.wait_group.finish();
+
+        const hct_buffer = try server.allocator.alignedAlloc(
+            u8,
+            @alignOf(HandleConnectionTask),
+            HandleConnectionTask.initBufferSize(server.read_buffer_size),
+        );
+        errdefer server.allocator.free(hct_buffer);
+
+        const new_hct = try HandleConnectionTask.initBufferAndReceiveHead(
+            hct_buffer,
+            server,
+            conn,
+        );
+        errdefer new_hct.http_server.connection.stream.close();
+
+        server.thread_pool.schedule(ThreadPool.Batch.from(&new_hct.task));
+    }
+};
+
+const HandleConnectionTask = struct {
+    task: ThreadPool.Task,
+    server: *Server,
+    http_server: std.http.Server,
+    request: std.http.Server.Request,
+
+    fn initBufferSize(read_buffer_size: usize) usize {
+        return @sizeOf(HandleConnectionTask) + read_buffer_size;
+    }
+
+    fn getFullBuffer(
+        read_buffer_size: usize,
+        hct: *HandleConnectionTask,
+    ) []align(@alignOf(HandleConnectionTask)) u8 {
+        const ptr: [*]align(@alignOf(HandleConnectionTask)) u8 = @ptrCast(hct);
+        return ptr[0..initBufferSize(read_buffer_size)];
+    }
+
+    fn getReadBuffer(
+        read_buffer_size: usize,
+        hct: *HandleConnectionTask,
+    ) []u8 {
+        return getFullBuffer(read_buffer_size, hct)[@sizeOf(HandleConnectionTask)..];
+    }
```

Comment:
would be good to have init methods first, then these helpers can come after - bit confusing navigating the code with this difference

Reply:
ce0c47543da8b6befabc76e4892e50154b0904ec